File: clients/fsharp/Debugging/DebugData.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Debugging

open AiCup22

/// Text
type DebugDataPlacedText = {
    /// Position
    Position: Model.Vec2;
    /// Text
    Text: string;
    /// Alignment, separate for x and y. From 0 to 1. 0.5 - center alignment
    Alignment: Model.Vec2;
    /// Size
    Size: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write PlacedText to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        this.Position.writeTo writer
        let textData : byte[] = System.Text.Encoding.UTF8.GetBytes this.Text
        writer.Write textData.Length
        writer.Write textData
        this.Alignment.writeTo writer
        writer.Write this.Size
        this.Color.writeTo writer
        ()

    /// Read PlacedText from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Text = reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString
        Alignment = Model.Vec2.readFrom reader;
        Size = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Circle
type DebugDataCircle = {
    /// Position of the center
    Position: Model.Vec2;
    /// Radius
    Radius: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Circle to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        this.Position.writeTo writer
        writer.Write this.Radius
        this.Color.writeTo writer
        ()

    /// Read Circle from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Circle with gradient fill
type DebugDataGradientCircle = {
    /// Position of the center
    Position: Model.Vec2;
    /// Radius
    Radius: double;
    /// Color of the center
    InnerColor: Debugging.Color;
    /// Color of the edge
    OuterColor: Debugging.Color;
} with

    /// Write GradientCircle to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        this.Position.writeTo writer
        writer.Write this.Radius
        this.InnerColor.writeTo writer
        this.OuterColor.writeTo writer
        ()

    /// Read GradientCircle from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        InnerColor = Debugging.Color.readFrom reader;
        OuterColor = Debugging.Color.readFrom reader;
    }

/// Ring
type DebugDataRing = {
    /// Position of the center
    Position: Model.Vec2;
    /// Radius
    Radius: double;
    /// Width
    Width: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Ring to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 3
        this.Position.writeTo writer
        writer.Write this.Radius
        writer.Write this.Width
        this.Color.writeTo writer
        ()

    /// Read Ring from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        Width = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Sector of a circle
type DebugDataPie = {
    /// Position of the center
    Position: Model.Vec2;
    /// Radius
    Radius: double;
    /// Start angle
    StartAngle: double;
    /// End angle
    EndAngle: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Pie to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 4
        this.Position.writeTo writer
        writer.Write this.Radius
        writer.Write this.StartAngle
        writer.Write this.EndAngle
        this.Color.writeTo writer
        ()

    /// Read Pie from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        StartAngle = reader.ReadDouble()
        EndAngle = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Arc
type DebugDataArc = {
    /// Position of the center
    Position: Model.Vec2;
    /// Radius
    Radius: double;
    /// Width
    Width: double;
    /// Start angle
    StartAngle: double;
    /// End angle
    EndAngle: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Arc to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 5
        this.Position.writeTo writer
        writer.Write this.Radius
        writer.Write this.Width
        writer.Write this.StartAngle
        writer.Write this.EndAngle
        this.Color.writeTo writer
        ()

    /// Read Arc from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        Width = reader.ReadDouble()
        StartAngle = reader.ReadDouble()
        EndAngle = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Rectancle
type DebugDataRect = {
    /// Bottom left position
    BottomLeft: Model.Vec2;
    /// Size
    Size: Model.Vec2;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Rect to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 6
        this.BottomLeft.writeTo writer
        this.Size.writeTo writer
        this.Color.writeTo writer
        ()

    /// Read Rect from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        BottomLeft = Model.Vec2.readFrom reader;
        Size = Model.Vec2.readFrom reader;
        Color = Debugging.Color.readFrom reader;
    }

/// Polygon (convex)
type DebugDataPolygon = {
    /// Positions of vertices in order
    Vertices: Model.Vec2[];
    /// Color
    Color: Debugging.Color;
} with

    /// Write Polygon to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 7
        writer.Write this.Vertices.Length
        this.Vertices |> Array.iter (fun value ->
            value.writeTo writer )
        this.Color.writeTo writer
        ()

    /// Read Polygon from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertices = [|for _ in 1 .. reader.ReadInt32() do
                       yield Model.Vec2.readFrom reader; |]
        Color = Debugging.Color.readFrom reader;
    }

/// Polygon with gradient fill
type DebugDataGradientPolygon = {
    /// List of vertices in order
    Vertices: Debugging.ColoredVertex[];
} with

    /// Write GradientPolygon to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 8
        writer.Write this.Vertices.Length
        this.Vertices |> Array.iter (fun value ->
            value.writeTo writer )
        ()

    /// Read GradientPolygon from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertices = [|for _ in 1 .. reader.ReadInt32() do
                       yield Debugging.ColoredVertex.readFrom reader; |]
    }

/// Segment
type DebugDataSegment = {
    /// Position of the first end
    FirstEnd: Model.Vec2;
    /// Position of the second end
    SecondEnd: Model.Vec2;
    /// Width
    Width: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write Segment to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 9
        this.FirstEnd.writeTo writer
        this.SecondEnd.writeTo writer
        writer.Write this.Width
        this.Color.writeTo writer
        ()

    /// Read Segment from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        FirstEnd = Model.Vec2.readFrom reader;
        SecondEnd = Model.Vec2.readFrom reader;
        Width = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Segment with gradient fill
type DebugDataGradientSegment = {
    /// Position of the first end
    FirstEnd: Model.Vec2;
    /// Color of the first end
    FirstColor: Debugging.Color;
    /// Position of the second end
    SecondEnd: Model.Vec2;
    /// Color of the second end
    SecondColor: Debugging.Color;
    /// Width
    Width: double;
} with

    /// Write GradientSegment to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 10
        this.FirstEnd.writeTo writer
        this.FirstColor.writeTo writer
        this.SecondEnd.writeTo writer
        this.SecondColor.writeTo writer
        writer.Write this.Width
        ()

    /// Read GradientSegment from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        FirstEnd = Model.Vec2.readFrom reader;
        FirstColor = Debugging.Color.readFrom reader;
        SecondEnd = Model.Vec2.readFrom reader;
        SecondColor = Debugging.Color.readFrom reader;
        Width = reader.ReadDouble()
    }

/// Poly line
type DebugDataPolyLine = {
    /// List of points in order
    Vertices: Model.Vec2[];
    /// Width
    Width: double;
    /// Color
    Color: Debugging.Color;
} with

    /// Write PolyLine to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 11
        writer.Write this.Vertices.Length
        this.Vertices |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.Width
        this.Color.writeTo writer
        ()

    /// Read PolyLine from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertices = [|for _ in 1 .. reader.ReadInt32() do
                       yield Model.Vec2.readFrom reader; |]
        Width = reader.ReadDouble()
        Color = Debugging.Color.readFrom reader;
    }

/// Poly line with gradient fill
type DebugDataGradientPolyLine = {
    /// List of points and colors in order
    Vertices: Debugging.ColoredVertex[];
    /// Width
    Width: double;
} with

    /// Write GradientPolyLine to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 12
        writer.Write this.Vertices.Length
        this.Vertices |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.Width
        ()

    /// Read GradientPolyLine from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertices = [|for _ in 1 .. reader.ReadInt32() do
                       yield Debugging.ColoredVertex.readFrom reader; |]
        Width = reader.ReadDouble()
    }

/// Data for debug rendering
type DebugData =
    /// Text
    | PlacedText of DebugDataPlacedText
    /// Circle
    | Circle of DebugDataCircle
    /// Circle with gradient fill
    | GradientCircle of DebugDataGradientCircle
    /// Ring
    | Ring of DebugDataRing
    /// Sector of a circle
    | Pie of DebugDataPie
    /// Arc
    | Arc of DebugDataArc
    /// Rectancle
    | Rect of DebugDataRect
    /// Polygon (convex)
    | Polygon of DebugDataPolygon
    /// Polygon with gradient fill
    | GradientPolygon of DebugDataGradientPolygon
    /// Segment
    | Segment of DebugDataSegment
    /// Segment with gradient fill
    | GradientSegment of DebugDataGradientSegment
    /// Poly line
    | PolyLine of DebugDataPolyLine
    /// Poly line with gradient fill
    | GradientPolyLine of DebugDataGradientPolyLine
    with

    /// Write DebugData to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | PlacedText value -> value.writeTo writer
            | Circle value -> value.writeTo writer
            | GradientCircle value -> value.writeTo writer
            | Ring value -> value.writeTo writer
            | Pie value -> value.writeTo writer
            | Arc value -> value.writeTo writer
            | Rect value -> value.writeTo writer
            | Polygon value -> value.writeTo writer
            | GradientPolygon value -> value.writeTo writer
            | Segment value -> value.writeTo writer
            | GradientSegment value -> value.writeTo writer
            | PolyLine value -> value.writeTo writer
            | GradientPolyLine value -> value.writeTo writer

    /// Read DebugData from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> PlacedText (DebugDataPlacedText.readFrom reader)
            | 1 -> Circle (DebugDataCircle.readFrom reader)
            | 2 -> GradientCircle (DebugDataGradientCircle.readFrom reader)
            | 3 -> Ring (DebugDataRing.readFrom reader)
            | 4 -> Pie (DebugDataPie.readFrom reader)
            | 5 -> Arc (DebugDataArc.readFrom reader)
            | 6 -> Rect (DebugDataRect.readFrom reader)
            | 7 -> Polygon (DebugDataPolygon.readFrom reader)
            | 8 -> GradientPolygon (DebugDataGradientPolygon.readFrom reader)
            | 9 -> Segment (DebugDataSegment.readFrom reader)
            | 10 -> GradientSegment (DebugDataGradientSegment.readFrom reader)
            | 11 -> PolyLine (DebugDataPolyLine.readFrom reader)
            | 12 -> GradientPolyLine (DebugDataGradientPolyLine.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)

File: clients/haskell/src/Debugging/DebugData.hs
module Debugging.DebugData where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Debugging.Color (Color)
import Debugging.ColoredVertex (ColoredVertex)
import Model.Vec2 (Vec2)

-- | Text
data DebugDataPlacedText = DebugDataPlacedText {
    -- | Position
    position :: Vec2,
    -- | Text
    text :: String,
    -- | Alignment, separate for x and y. From 0 to 1. 0.5 - center alignment
    alignment :: Vec2,
    -- | Size
    size :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataPlacedText where
    read = do
        position <- Trans.read
        text <- Trans.read
        alignment <- Trans.read
        size <- Trans.read
        color <- Trans.read
        return DebugDataPlacedText {
            position,
            text,
            alignment,
            size,
            color }
    
    write DebugDataPlacedText {
        position,
        text,
        alignment,
        size,
        color } = do
            Trans.write position
            Trans.write text
            Trans.write alignment
            Trans.write size
            Trans.write color

-- | Circle
data DebugDataCircle = DebugDataCircle {
    -- | Position of the center
    position :: Vec2,
    -- | Radius
    radius :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataCircle where
    read = do
        position <- Trans.read
        radius <- Trans.read
        color <- Trans.read
        return DebugDataCircle {
            position,
            radius,
            color }
    
    write DebugDataCircle {
        position,
        radius,
        color } = do
            Trans.write position
            Trans.write radius
            Trans.write color

-- | Circle with gradient fill
data DebugDataGradientCircle = DebugDataGradientCircle {
    -- | Position of the center
    position :: Vec2,
    -- | Radius
    radius :: Double,
    -- | Color of the center
    innerColor :: Color,
    -- | Color of the edge
    outerColor :: Color }
    deriving Show

instance Trans DebugDataGradientCircle where
    read = do
        position <- Trans.read
        radius <- Trans.read
        innerColor <- Trans.read
        outerColor <- Trans.read
        return DebugDataGradientCircle {
            position,
            radius,
            innerColor,
            outerColor }
    
    write DebugDataGradientCircle {
        position,
        radius,
        innerColor,
        outerColor } = do
            Trans.write position
            Trans.write radius
            Trans.write innerColor
            Trans.write outerColor

-- | Ring
data DebugDataRing = DebugDataRing {
    -- | Position of the center
    position :: Vec2,
    -- | Radius
    radius :: Double,
    -- | Width
    width :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataRing where
    read = do
        position <- Trans.read
        radius <- Trans.read
        width <- Trans.read
        color <- Trans.read
        return DebugDataRing {
            position,
            radius,
            width,
            color }
    
    write DebugDataRing {
        position,
        radius,
        width,
        color } = do
            Trans.write position
            Trans.write radius
            Trans.write width
            Trans.write color

-- | Sector of a circle
data DebugDataPie = DebugDataPie {
    -- | Position of the center
    position :: Vec2,
    -- | Radius
    radius :: Double,
    -- | Start angle
    startAngle :: Double,
    -- | End angle
    endAngle :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataPie where
    read = do
        position <- Trans.read
        radius <- Trans.read
        startAngle <- Trans.read
        endAngle <- Trans.read
        color <- Trans.read
        return DebugDataPie {
            position,
            radius,
            startAngle,
            endAngle,
            color }
    
    write DebugDataPie {
        position,
        radius,
        startAngle,
        endAngle,
        color } = do
            Trans.write position
            Trans.write radius
            Trans.write startAngle
            Trans.write endAngle
            Trans.write color

-- | Arc
data DebugDataArc = DebugDataArc {
    -- | Position of the center
    position :: Vec2,
    -- | Radius
    radius :: Double,
    -- | Width
    width :: Double,
    -- | Start angle
    startAngle :: Double,
    -- | End angle
    endAngle :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataArc where
    read = do
        position <- Trans.read
        radius <- Trans.read
        width <- Trans.read
        startAngle <- Trans.read
        endAngle <- Trans.read
        color <- Trans.read
        return DebugDataArc {
            position,
            radius,
            width,
            startAngle,
            endAngle,
            color }
    
    write DebugDataArc {
        position,
        radius,
        width,
        startAngle,
        endAngle,
        color } = do
            Trans.write position
            Trans.write radius
            Trans.write width
            Trans.write startAngle
            Trans.write endAngle
            Trans.write color

-- | Rectancle
data DebugDataRect = DebugDataRect {
    -- | Bottom left position
    bottomLeft :: Vec2,
    -- | Size
    size :: Vec2,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataRect where
    read = do
        bottomLeft <- Trans.read
        size <- Trans.read
        color <- Trans.read
        return DebugDataRect {
            bottomLeft,
            size,
            color }
    
    write DebugDataRect {
        bottomLeft,
        size,
        color } = do
            Trans.write bottomLeft
            Trans.write size
            Trans.write color

-- | Polygon (convex)
data DebugDataPolygon = DebugDataPolygon {
    -- | Positions of vertices in order
    vertices :: [Vec2],
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataPolygon where
    read = do
        vertices <- Trans.read
        color <- Trans.read
        return DebugDataPolygon {
            vertices,
            color }
    
    write DebugDataPolygon {
        vertices,
        color } = do
            Trans.write vertices
            Trans.write color

-- | Polygon with gradient fill
data DebugDataGradientPolygon = DebugDataGradientPolygon {
    -- | List of vertices in order
    vertices :: [ColoredVertex] }
    deriving Show

instance Trans DebugDataGradientPolygon where
    read = do
        vertices <- Trans.read
        return DebugDataGradientPolygon {
            vertices }
    
    write DebugDataGradientPolygon {
        vertices } = do
            Trans.write vertices

-- | Segment
data DebugDataSegment = DebugDataSegment {
    -- | Position of the first end
    firstEnd :: Vec2,
    -- | Position of the second end
    secondEnd :: Vec2,
    -- | Width
    width :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataSegment where
    read = do
        firstEnd <- Trans.read
        secondEnd <- Trans.read
        width <- Trans.read
        color <- Trans.read
        return DebugDataSegment {
            firstEnd,
            secondEnd,
            width,
            color }
    
    write DebugDataSegment {
        firstEnd,
        secondEnd,
        width,
        color } = do
            Trans.write firstEnd
            Trans.write secondEnd
            Trans.write width
            Trans.write color

-- | Segment with gradient fill
data DebugDataGradientSegment = DebugDataGradientSegment {
    -- | Position of the first end
    firstEnd :: Vec2,
    -- | Color of the first end
    firstColor :: Color,
    -- | Position of the second end
    secondEnd :: Vec2,
    -- | Color of the second end
    secondColor :: Color,
    -- | Width
    width :: Double }
    deriving Show

instance Trans DebugDataGradientSegment where
    read = do
        firstEnd <- Trans.read
        firstColor <- Trans.read
        secondEnd <- Trans.read
        secondColor <- Trans.read
        width <- Trans.read
        return DebugDataGradientSegment {
            firstEnd,
            firstColor,
            secondEnd,
            secondColor,
            width }
    
    write DebugDataGradientSegment {
        firstEnd,
        firstColor,
        secondEnd,
        secondColor,
        width } = do
            Trans.write firstEnd
            Trans.write firstColor
            Trans.write secondEnd
            Trans.write secondColor
            Trans.write width

-- | Poly line
data DebugDataPolyLine = DebugDataPolyLine {
    -- | List of points in order
    vertices :: [Vec2],
    -- | Width
    width :: Double,
    -- | Color
    color :: Color }
    deriving Show

instance Trans DebugDataPolyLine where
    read = do
        vertices <- Trans.read
        width <- Trans.read
        color <- Trans.read
        return DebugDataPolyLine {
            vertices,
            width,
            color }
    
    write DebugDataPolyLine {
        vertices,
        width,
        color } = do
            Trans.write vertices
            Trans.write width
            Trans.write color

-- | Poly line with gradient fill
data DebugDataGradientPolyLine = DebugDataGradientPolyLine {
    -- | List of points and colors in order
    vertices :: [ColoredVertex],
    -- | Width
    width :: Double }
    deriving Show

instance Trans DebugDataGradientPolyLine where
    read = do
        vertices <- Trans.read
        width <- Trans.read
        return DebugDataGradientPolyLine {
            vertices,
            width }
    
    write DebugDataGradientPolyLine {
        vertices,
        width } = do
            Trans.write vertices
            Trans.write width

-- | Data for debug rendering
data DebugData
    -- | Text
    = PlacedText DebugDataPlacedText
    -- | Circle
    | Circle DebugDataCircle
    -- | Circle with gradient fill
    | GradientCircle DebugDataGradientCircle
    -- | Ring
    | Ring DebugDataRing
    -- | Sector of a circle
    | Pie DebugDataPie
    -- | Arc
    | Arc DebugDataArc
    -- | Rectancle
    | Rect DebugDataRect
    -- | Polygon (convex)
    | Polygon DebugDataPolygon
    -- | Polygon with gradient fill
    | GradientPolygon DebugDataGradientPolygon
    -- | Segment
    | Segment DebugDataSegment
    -- | Segment with gradient fill
    | GradientSegment DebugDataGradientSegment
    -- | Poly line
    | PolyLine DebugDataPolyLine
    -- | Poly line with gradient fill
    | GradientPolyLine DebugDataGradientPolyLine
    deriving Show

instance Trans DebugData where
    read = do
        tag :: Int32 <- Trans.read
        case tag of
            0 -> PlacedText <$> Trans.read
            1 -> Circle <$> Trans.read
            2 -> GradientCircle <$> Trans.read
            3 -> Ring <$> Trans.read
            4 -> Pie <$> Trans.read
            5 -> Arc <$> Trans.read
            6 -> Rect <$> Trans.read
            7 -> Polygon <$> Trans.read
            8 -> GradientPolygon <$> Trans.read
            9 -> Segment <$> Trans.read
            10 -> GradientSegment <$> Trans.read
            11 -> PolyLine <$> Trans.read
            12 -> GradientPolyLine <$> Trans.read
    
    write (PlacedText value) = do
        Trans.write (0 :: Int32)
        Trans.write value
    write (Circle value) = do
        Trans.write (1 :: Int32)
        Trans.write value
    write (GradientCircle value) = do
        Trans.write (2 :: Int32)
        Trans.write value
    write (Ring value) = do
        Trans.write (3 :: Int32)
        Trans.write value
    write (Pie value) = do
        Trans.write (4 :: Int32)
        Trans.write value
    write (Arc value) = do
        Trans.write (5 :: Int32)
        Trans.write value
    write (Rect value) = do
        Trans.write (6 :: Int32)
        Trans.write value
    write (Polygon value) = do
        Trans.write (7 :: Int32)
        Trans.write value
    write (GradientPolygon value) = do
        Trans.write (8 :: Int32)
        Trans.write value
    write (Segment value) = do
        Trans.write (9 :: Int32)
        Trans.write value
    write (GradientSegment value) = do
        Trans.write (10 :: Int32)
        Trans.write value
    write (PolyLine value) = do
        Trans.write (11 :: Int32)
        Trans.write value
    write (GradientPolyLine value) = do
        Trans.write (12 :: Int32)
        Trans.write value